knative-operator/pkg/controller/knativeserving/knativeserving_controller.go

```diff
@@ -224,49 +233,97 @@ func (r *ReconcileKnativeServing) ensureFinalizers(instance *servingv1alpha1.Kna
 // create the configmap to be injected with custom certs
 func (r *ReconcileKnativeServing) ensureCustomCertsConfigMap(instance *servingv1alpha1.KnativeServing) error {
 	const (
-		// Using both the serviceCA annotation and the trustedCA label is safe. The serviceCA reconciler
-		// will only reconcile once and then short-circuit on an existing key, see https://git.io/JvNSr.
-		// The trustedCA label reconciles in an additive way, see https://git.io/JvNSV.
-
 		// Docs: https://github.com/openshift/service-ca-operator
-		serviceCAKey = "service.alpha.openshift.io/inject-cabundle"
+		serviceCAKey     = "service.alpha.openshift.io/inject-cabundle"
+		serviceCAContent = "service-ca.crt"
 		// Docs: https://docs.openshift.com/container-platform/4.3/networking/configuring-a-custom-pki.html#certificate-injection-using-operators_configuring-a-custom-pki
-		trustedCAKey = "config.openshift.io/inject-trusted-cabundle"
+		trustedCAKey     = "config.openshift.io/inject-trusted-cabundle"
+		trustedCAContent = "tls-ca-bundle.pem"
 	)
 
 	certs := instance.Spec.ControllerCustomCerts
+
+	// If the user doesn't specify anything else, this is set by the webhook/controller defaulter to
+	// cause us to automatically pull in the relevant ConfigMaps from the cluster. The user needs
+	// to specifically opt-out of this today by specifying an empty Name and ConfigMap explicitly.
 	if certs.Type != "ConfigMap" || certs.Name == "" {
 		return nil
 	}
-	cm := &corev1.ConfigMap{}
+
+	serviceCACM, err := r.reconcileConfigMap(instance, certs.Name+"-service-ca", map[string]string{serviceCAKey: "true"}, nil, nil)
+	if err != nil {
+		return fmt.Errorf("error reconciling serviceCACM: %w", err)
+	}
+	trustedCACM, err := r.reconcileConfigMap(instance, certs.Name+"-trusted-ca", nil, map[string]string{trustedCAKey: "true"}, nil)
+	if err != nil {
+		return fmt.Errorf("error reconciling serviceCACM: %w", err)
+	}
+
+	combinedContents := make(map[string]string, len(serviceCACM.Data)+len(trustedCACM.Data))
```

Comment:
Just so I understand the flow here, the first time this happens we'll effectively create an empty combined ConfigMap.  And as the underlying ConfigMaps get their certs set by OpenShift, we'll then start combining the ConfigMaps. Does the Knative Serving controller automatically watch for and pick up any changes to the certs in its mounted ConfigMap volume?

Reply:
That's correct and no the controller actually does not watch this currently. I was poking upstream about the possibility to do that.

FWIW: This is a hole on master as well. The timing usually comes around nicely for us so far. I did some refactoring today too though to implement a "kick" from our operator. I plan to set an env var on the deployment with the resourceVersion of the configMap effectively. That should also give us cert rotation in general.

Reply:
Let's open a JIRA issue if there's not one already to track the kick/env var/cert rotation work. Given that, this looks good.

Reply:
wdm!